File: tests/cohorts.py
# A set of tools for manipulating test braids

import json, os, unittest
from copy import copy
from collections import deque, OrderedDict

def makedag(parents, description=None):
    """ Make a DAG object which caches the children, geneses, tips, cohorts, and highest work path. """
    dag = {}
    dag["description"] = description
    dag["parents"] = parents
    dag["children"] = reverse(parents)
    dag["cohorts"] = cohorts(parents)
    dag["geneses"] = geneses(parents)
    dag["tips"] = geneses(dag["children"])
    dag["highestworkpath"] = highest_work_path(parents, dag["children"])
    return dag

def geneses(parents):
    """ Given a dict of {bead: {parents}}, return the set of beads which have no parents. """
    geneses = set()
    for b,p in parents.items():
        if not p: geneses.add(b)
    return geneses

def tips(parents):
    """ Given a dict of {bead: {parents}}, return the set of beads which have no children. """
    return geneses(reverse(parents))

def reverse(parents):
    """ Given a dict of {bead: {parents}}, compute the corresponding {bead: {children}} (or
        vice-versa).
    """
    children = {}
    for bead, bparents in parents.items():
        if bead not in children:
            children[bead] = set()

        for p in bparents:
            if p not in children:
                children[p] = set()
            children[p].add(bead)
    return children

def generation(beads, children):
    """ Given a set of <beads>, compute the set of all children of all {beads}. """
    if not beads: return set()
    retval = set()
    for b in beads:
        retval |= children[b]
    return retval

def getallancestors(b, parents, ancestors):
    """ Gets all ancestors for a bead <b>, filling in ancestors of
        any other ancestors encountered, using a recursive
        algorithm.  Assumes b not in parents and b not in ancestors.
    """
    ancestors[b] = set(copy(parents[b]))
    for p in parents[b]:
        if p not in ancestors: getallancestors(p, parents, ancestors)
        ancestors[b].update(ancestors[p])

def cohorts(parents, initial_cohort=None):
    """ Given the seed of the next cohort (which is the set of beads one step older, in the next
        cohort), build an ancestor/descendant set for each visited bead.  A new cohort is
        formed if we encounter a set of beads, stepping in the descendant direction, for which
        *all* beads in this cohort are ancestors of the first generation of beads in the next
        cohort.
    """
    dag          = {"parents": parents, "children": reverse(parents), "tips": tips(parents)}
    cohort       = initial_cohort or geneses(parents)
    oldcohort    = set()
    head         = copy(cohort)
    tail         = copy(cohort)
    while True :
        ancestors = {h: set() for h in head}            # Don't let head have ancestors to stop ancestor iteration
        cohort    = copy(head)                          # Initial cohort is the head

        while True:                                     # DFS search
            # Calculate new tail
            if not head: return                         # StopIteration and return
            else:
                for b in cohort-oldcohort:
                    tail |= dag["children"][b]          # Add the next generation to the tail
                tail |= cohort ^ oldcohort              # Add any beads in the oldcohort but not in the new cohort
                if cohort & dag["tips"]:                # If there are any tips in cohort, add remaining tips to tail:
                    tail |= dag["tips"]-cohort
                else:
                    tail -= cohort                      # If there are no tips in the cohort subtract off the cohort

            oldcohort = copy(cohort)                    # Copy so we can tell that a new tail hasn't changed anything
                                                        # Changing the tail but not cohort may find new ancestors

            # Calculate ancestors
            for t in tail:                              # Find all ancestors of all beads in the tail
                if t not in ancestors:
                    # 50.6% of CPU time
                    getallancestors(t, dag["parents"], ancestors)

            # Calculate cohort
            cohort = set()
            for a in ancestors: cohort |= ancestors[a]  # Union all ancestors with the cohort

            # Check termination cases
            if dag["tips"] <= cohort:                          # Cohort contains all tips
                head = set()                            # StopIteration and return
                break                                   # and yield the current cohort
            elif cohort and all(ancestors[t] == cohort for t in tail): # Standard cohort case
                head = copy(tail)                       # Head of next cohort is the tail from the previous iteration
                break                                   # Yield successful cohort
            elif cohort == oldcohort:                   # Cohort hasn't changed, we may be looping
                if dag["tips"] <= tail:                        # Tail contains all tips but we didn't form a cohort
                    head = set()                        # We also need cohort == oldcohort (yes)
                    cohort |= tail
                    tail = set()
                    break                               # Yield cohort+tail
                else: cohort.update(tail)               # Seems like I'll be doing 2 loops for non-cohort updates
                                                        # This breaks if I omit the cohort == oldcohort condition.
                                                        #elif not tips & tail and not cohort & tips: (doesn't work)
        oldcohort = set()
        yield cohort

def cohort_head(cohort, parents, children):
    """ Given a cohort as a set of beads, compute its head. """
    return cohort_headtail(cohort, parents, children)

def cohort_tail(cohort, parents, children):
    """ Given a cohort as a set of beads, compute its tail. If the tail intersects the tips, return all tips. """
    return cohort_headtail(cohort, children, parents)

def cohort_headtail(cohort, parents, children):
    tail = generation(generation(cohort, parents) - cohort, children)
    tips = geneses(parents)
    if not tail or any({t in tips for t in tail}):
        return tips
    return tail

def highest_work_path(parents, children):
    """ Find the highest work path by adding the work of each parent
        bead using BFS.  We only need to do this within cohorts, since a
        cohort has all the same ancestors and descendants. This assumes
        all beads have the same work.
    """
    hwpath = []
    work = 1 # work per bead. Update this if if a DAA is used and different beads have different work
    for c in cohorts(parents):
        head = cohort_head(c, parents, children) # Youngest set of beads in the cohort
        tail = cohort_tail(c, parents, children) # Oldest set of beads in the cohort
        if head == tail:
            hwpath.append(next(iter(c)))
            continue
        queue = deque(set.union(*[children[h] for h in head]))
        weights = {h: work for h in head}

        # build weights dict
        while queue:
           b = queue.pop()
           cparents = generation({b}, parents)
           if b not in weights:
               missingparents = [p for p in cparents if p not in weights]
               if missingparents:
                   queue.extend(missingparents)
                   continue
               else:
                   weights[b] = work + sum([weights[p] for p in cparents])
                   queue.extendleft(generation({b}, children))

        # Follow the highest weights through the DAG from tail to head and build hwpath
        chwpath = [max(tail, key=lambda x: weights[x])]
        while chwpath[-1] not in head:
            chwpath.append(max(generation({chwpath[-1]}, parents), key=lambda x: weights[x]))
        hwpath.extend(reversed(chwpath))
    return hwpath

def load_braid(filename):
    """ Load a JSON file containing a braid.
    """
    with open(filename) as json_data:
        d = json.load(json_data)
        dag = {}
        dag["description"] = d["description"]
        dag["parents"] = {int(k): set(v) for k,v in d["parents"].items()}
        dag["children"] = {int(k): set(v) for k,v in d["children"].items()}
        dag["cohorts"] = [set(map(int,c)) for c in d["cohorts"]]
        dag["highestworkpath"] = d["highestworkpath"]
        dag["tips"] = set(d["tips"])
        dag["geneses"] = set(d["geneses"])

        json_data.close()
        return dag

def save_braid(parents, filename, description=None):
    """ Save a JSON file containing a braid. It should contain the keys "description", "parents",
        "cohorts", and "workpath"
    """
    with open(filename, 'w') as file:
        dag = makedag(parents)
        result = OrderedDict([
            ("description", description),
            ("parents", {k: list(v) for k,v in dag["parents"].items()}),
            ("children", {k: list(v) for k,v in dag["children"].items()}),
            ("geneses", list(dag["geneses"])),
            ("tips", list(dag["tips"])),
            ("cohorts", [sorted(list(map(int, c))) for c in dag["cohorts"]]),
            ("highestworkpath", dag["highestworkpath"])
        ])
        file.write(json.dumps(result, sort_keys=False, indent=4))
        file.close()

class TestCohortMethods(unittest.TestCase):
    parents1 = {0:set(), 1:{0}, 2:{1}, 3:{2}} # blockchain
    parents2 = {0:set(), 1:set(), 2:{1}, 3:{1}}
    parents3 = {0:set(), 1:set(), 2:set(), 3:{1}, 4:{0}}
    parents4 = {0:set(), 1:set(), 2:set(), 3:{0,1,2}, 4:{0,1,2}, 5:{0,1,2}}
    diamond  = {0:set(), 1:{0}, 2:{0}, 3:{1,2}, 4:{3}}

    def test_geneses1(self):
        self.assertEqual(geneses(self.parents1), {0})

    def test_geneses2(self):
        self.assertEqual(geneses(self.parents2), {0, 1})

    def test_geneses3(self):
        self.assertEqual(geneses(self.parents3), {0, 1, 2})

    def test_geneses_files(self):
        for filename in sorted([filename for filename in os.listdir() if filename.endswith(".braid")]):
            dag = load_braid(filename)
            self.assertEqual(geneses(dag["parents"]), {0})

    def test_tips1(self):
        parents1 = {0:set(), 1:[0], 2:[1], 3:[2]}
        self.assertEqual(tips(parents1), {3})

    def test_tips2(self):
        parents2 = {0:set(), 1:[0], 2:[1], 3:[1]}
        self.assertEqual(tips(parents2), {2,3})

    def test_tips3(self):
        parents3 = {0:set(), 1:set(), 2:set(), 3:[0,1,2], 4:[0,1,2], 5:[0,1,2]}
        self.assertEqual(tips(parents3), {3,4,5})

    def test_reverse(self):
        self.assertEqual(reverse(self.parents4),
                         {0: {3, 4, 5}, 1: {3, 4, 5}, 2: {3, 4, 5}, 3: set(), 4: set(), 5: set()})

    def test_cohorts(self):
        """ Test cohorts for a trivial blockchain-like braid. """
        self.assertEqual(list(cohorts(self.parents1)), [{0}, {1}, {2}, {3}])

    def test_cohorts_files(self):
        """ Test an assortment of example *.braid files. """
        for filename in sorted([filename for filename in os.listdir() if filename.endswith(".braid")]):
            dag = load_braid(filename)
            self.assertEqual(list(cohorts(dag["parents"])), dag["cohorts"])

    def test_cohorts_reversed_files(self):
        """ Does it find the same cohorts in the forward and backward directions? """
        for filename in sorted([filename for filename in os.listdir() if filename.endswith(".braid")]):
            dag = load_braid(filename)
            p = reverse(dag["parents"])
            c = dag["cohorts"]
            c.reverse()
            self.assertEqual(list(cohorts(p)), c)

    def test_highest_work_path(self):
        self.assertEqual(highest_work_path(self.parents1, reverse(self.parents1)), [0,1,2,3])

    def test_higest_work_path_files(self):
        for filename in sorted([filename for filename in os.listdir() if filename.endswith(".braid")]):
            dag = load_braid(filename)
            self.assertEqual(highest_work_path(dag["parents"], dag["children"]), dag["highestworkpath"])


if __name__ == "__main__":
    unittest.main()

File: tests/src/cohorts.rs
// src/cohorts.rs
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet, VecDeque};
use std::fs::File;
use std::io::{Read, Write};
use std::path::Path;

// Type aliases to make the code more readable
pub type Bead = i32;
pub type Parents = HashMap<Bead, HashSet<Bead>>;
pub type Children = HashMap<Bead, HashSet<Bead>>;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Dag {
    pub description: Option<String>,
    #[serde(deserialize_with = "deserialize_parents")]
    pub parents: Parents,
    #[serde(deserialize_with = "deserialize_parents")]
    pub children: Children,
    #[serde(deserialize_with = "deserialize_cohorts")]
    pub cohorts: Vec<HashSet<Bead>>,
    #[serde(deserialize_with = "deserialize_set")]
    pub geneses: HashSet<Bead>,
    #[serde(deserialize_with = "deserialize_set")]
    pub tips: HashSet<Bead>,
    pub highest_work_path: Vec<Bead>,
}

impl Dag {
    pub fn new(parents: Parents, description: Option<String>) -> Self {
        let children = reverse(&parents);
        let cohorts = get_cohorts(&parents).collect();
        let geneses = get_geneses(&parents);
        let tips = get_geneses(&children);
        let highest_work_path = highest_work_path(&parents, &children);

        Dag {
            description,
            parents,
            children,
            cohorts,
            geneses,
            tips,
            highest_work_path,
        }
    }

    pub fn load(path: impl AsRef<Path>) -> std::io::Result<Self> {
        let mut file = File::open(path)?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        Ok(serde_json::from_str(&contents)?)
    }

    pub fn save(&self, path: impl AsRef<Path>) -> std::io::Result<()> {
        let json = serde_json::to_string_pretty(&self)?;
        let mut file = File::create(path)?;
        file.write_all(json.as_bytes())?;
        Ok(())
    }
}

struct CohortsIterator {
    dag: HashMap<String, HashMap<Bead, HashSet<Bead>>>,
    cohort: HashSet<Bead>,
    old_cohort: HashSet<Bead>,
    head: HashSet<Bead>,
    tail: HashSet<Bead>,
}

// Custom deserializers for JSON format compatibility
fn deserialize_parents<'de, D>(deserializer: D) -> Result<HashMap<Bead, HashSet<Bead>>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let map: HashMap<String, Vec<Bead>> = HashMap::deserialize(deserializer)?;
    Ok(map
        .into_iter()
        .map(|(k, v)| (k.parse::<Bead>().unwrap(), v.into_iter().collect()))
        .collect())
}

fn deserialize_cohorts<'de, D>(deserializer: D) -> Result<Vec<HashSet<Bead>>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let vec: Vec<Vec<Bead>> = Vec::deserialize(deserializer)?;
    Ok(vec.into_iter().map(|v| v.into_iter().collect()).collect())
}

fn deserialize_set<'de, D>(deserializer: D) -> Result<HashSet<Bead>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let vec: Vec<Bead> = Vec::deserialize(deserializer)?;
    Ok(vec.into_iter().collect())
}

pub fn get_geneses(parents: &Parents) -> HashSet<Bead> {
    parents
        .iter()
        .filter(|(_, p)| p.is_empty())
        .map(|(b, _)| *b)
        .collect()
}

pub fn get_tips(parents: &Parents) -> HashSet<Bead> {
    get_geneses(&reverse(parents))
}

pub fn reverse(parents: &Parents) -> Children {
    let mut children: Children = HashMap::new();

    for (&bead, bparents) in parents.iter() {
        children.entry(bead).or_insert_with(HashSet::new);

        for &p in bparents {
            children.entry(p).or_insert_with(HashSet::new).insert(bead);
        }
    }
    children
}

fn generation(beads: &HashSet<Bead>, map: &HashMap<Bead, HashSet<Bead>>) -> HashSet<Bead> {
    if beads.is_empty() {
        return HashSet::new();
    }
    let mut result = HashSet::new();
    for &b in beads {
        if let Some(set) = map.get(&b) {
            result.extend(set);
        }
    }
    result
}

fn get_all_ancestors(b: Bead, parents: &Parents, ancestors: &mut HashMap<Bead, HashSet<Bead>>) {
    if !ancestors.contains_key(&b) {
        // Get parent beads first
        let parent_beads = parents.get(&b).cloned().unwrap_or_default();

        // Create initial ancestor set with just parents
        let mut b_ancestors = parent_beads.clone();

        // Process each parent's ancestors
        for p in parent_beads {
            if !ancestors.contains_key(&p) {
                get_all_ancestors(p, parents, ancestors);
            }
            if let Some(p_ancestors) = ancestors.get(&p) {
                b_ancestors.extend(p_ancestors);
            }
        }

        // Insert the complete ancestor set
        ancestors.insert(b, b_ancestors);
    }
}

impl CohortsIterator {
    pub fn new(parents: &Parents) -> Self {
        let children = reverse(parents);
        let initial_cohort = get_geneses(parents);

        let mut dag = HashMap::new();
        dag.insert("parents".to_string(), parents.clone());
        dag.insert("children".to_string(), children);

        CohortsIterator {
            dag,
            cohort: initial_cohort.clone(),
            old_cohort: HashSet::new(),
            head: initial_cohort.clone(),
            tail: initial_cohort,
        }
    }
}

impl Iterator for CohortsIterator {
    type Item = HashSet<Bead>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.head.is_empty() {
            return None;
        }

        let parents = self.dag.get("parents").unwrap();
        let children = self.dag.get("children").unwrap();
        let tips = get_tips(parents);

        let mut ancestors: HashMap<Bead, HashSet<Bead>> =
            self.head.iter().map(|&h| (h, HashSet::new())).collect();

        self.cohort = self.head.clone();

        loop {
            // Calculate new tail
            let mut new_tail = HashSet::new();
            for b in self.cohort.difference(&self.old_cohort) {
                if let Some(b_children) = children.get(b) {
                    new_tail.extend(b_children.iter().copied());
                }
            }

            // Add any beads in the oldcohort but not in the new cohort
            new_tail.extend(self.cohort.symmetric_difference(&self.old_cohort));

            // Handle tips in cohort
            if !self.cohort.is_disjoint(&tips) {
                // If there are tips in cohort, add remaining tips
                new_tail.extend(tips.difference(&self.cohort));
            } else {
                // If there are no tips in the cohort subtract off the cohort
                new_tail = &new_tail - &self.cohort;
            }

            self.tail = new_tail;
            self.old_cohort = self.cohort.clone();

            // Calculate ancestors
            for &t in &self.tail {
                if !ancestors.contains_key(&t) {
                    get_all_ancestors(t, parents, &mut ancestors);
                }
            }

            // Calculate cohort
            let mut new_cohort = HashSet::new();
            for (_, ancestor_set) in &ancestors {
                new_cohort.extend(ancestor_set);
            }
            self.cohort = new_cohort;

            // Check termination cases
            if tips.is_subset(&self.cohort) {
                // Cohort contains all tips
                self.head.clear();
                break;
            } else if !self.cohort.is_empty()
                && self
                    .tail
                    .iter()
                    .all(|t| ancestors.get(t).map_or(false, |a| a == &self.cohort))
            {
                // Standard cohort case
                self.head = self.tail.clone();
                break;
            } else if self.cohort == self.old_cohort {
                // Cohort hasn't changed, we may be looping
                if tips.is_subset(&self.tail) {
                    self.head.clear();
                    self.cohort.extend(&self.tail);
                    self.tail.clear();
                    break;
                } else {
                    self.cohort.extend(&self.tail);
                }
            }
        }

        self.old_cohort.clear();
        Some(self.cohort.clone())
    }
}

pub fn get_cohorts(parents: &Parents) -> impl Iterator<Item = HashSet<Bead>> + '_ {
    let children = reverse(parents);
    let initial_cohort = get_geneses(parents);
    let dag = DagInfo {
        parents: parents.clone(),
        children: children.clone(),
        tips: get_geneses(&children),
    };

    std::iter::from_fn(move || {
        static mut COHORT: Option<HashSet<Bead>> = None;
        static mut HEAD: Option<HashSet<Bead>> = None;

        // Initialize state if first iteration
        unsafe {
            if COHORT.is_none() {
                COHORT = Some(initial_cohort.clone());
                HEAD = Some(initial_cohort.clone());
            }

            let head = HEAD.as_mut()?;
            if head.is_empty() {
                return None;
            }

            let mut tail = head.clone();
            let mut cohort = head.clone();
            let mut old_cohort = HashSet::new();

            while !head.is_empty() {
                let mut ancestors: HashMap<Bead, HashSet<Bead>> =
                    head.iter().map(|&h| (h, HashSet::new())).collect();

                cohort = head.clone();

                // Inner loop (DFS search)
                loop {
                    // Calculate new tail
                    for b in cohort.difference(&old_cohort) {
                        if let Some(children_set) = dag.children.get(b) {
                            tail.extend(children_set);
                        }
                    }
                    tail.extend(cohort.symmetric_difference(&old_cohort));

                    if !cohort.is_disjoint(&dag.tips) {
                        tail.extend(dag.tips.difference(&cohort));
                    } else {
                        tail = tail.difference(&cohort).copied().collect();
                    }

                    old_cohort = cohort.clone();

                    // Calculate ancestors
                    for &t in &tail {
                        if !ancestors.contains_key(&t) {
                            get_all_ancestors(t, &dag.parents, &mut ancestors);
                        }
                    }

                    // Calculate cohort
                    let mut new_cohort = HashSet::new();
                    for ancestors_set in ancestors.values() {
                        new_cohort.extend(ancestors_set);
                    }
                    cohort = new_cohort;

                    // Check termination cases
                    if dag.tips.is_subset(&cohort) {
                        head.clear();
                        break;
                    } else if !cohort.is_empty()
                        && tail
                            .iter()
                            .all(|t| ancestors.get(t).map_or(false, |a| *a == cohort))
                    {
                        *head = tail.clone();
                        break;
                    } else if cohort == old_cohort {
                        if dag.tips.is_subset(&tail) {
                            head.clear();
                            cohort.extend(&tail);
                            break;
                        } else {
                            cohort.extend(&tail);
                        }
                    }
                }

                if head.is_empty() {
                    COHORT = None;
                    HEAD = None;
                }

                return Some(cohort);
            }

            None
        }
    })
}

pub fn highest_work_path(parents: &Parents, children: &Children) -> Vec<Bead> {
    let mut hwpath = Vec::new();
    let work = 1; // work per bead

    // Get cohorts iterator once and collect it so we process each cohort exactly once
    let cohorts: Vec<_> = get_cohorts(parents).collect();

    for c in cohorts {
        let head = cohort_head(&c, parents, children);
        let tail = cohort_tail(&c, parents, children);

        // Special case: if head equals tail, just take one bead from the cohort
        if head == tail {
            if let Some(&bead) = c.iter().next() {
                hwpath.push(bead);
                continue;
            }
        }

        // Initialize queue with union of all children of head beads
        let mut queue = VecDeque::new();
        for h in &head {
            if let Some(h_children) = children.get(h) {
                queue.extend(h_children.iter().copied());
            }
        }

        // Initialize weights for head beads
        let mut weights: HashMap<Bead, i32> = head.iter().map(|&h| (h, work)).collect();

        // Build weights dict
        while let Some(b) = queue.pop_front() {
            if !weights.contains_key(&b) {
                let cparents = generation(&HashSet::from([b]), parents);
                let missing_parents: Vec<_> = cparents
                    .iter()
                    .filter(|&p| !weights.contains_key(p))
                    .copied()
                    .collect();

                if !missing_parents.is_empty() {
                    // If any parents are missing weights, process them first
                    queue.extend(missing_parents);
                    continue;
                } else {
                    // Calculate weight for current bead
                    let weight = work + cparents.iter().map(|p| weights[p]).sum::<i32>();
                    weights.insert(b, weight);

                    // Add children to left of queue (equivalent to extendleft in Python)
                    if let Some(b_children) = children.get(&b) {
                        for &child in b_children {
                            queue.push_front(child);
                        }
                    }
                }
            }
        }

        // Follow highest weights from tail to head
        if !tail.is_empty() {
            // Find starting bead in tail with highest weight
            let mut current = *tail
                .iter()
                .max_by_key(|&&x| weights.get(&x).unwrap_or(&0))
                .unwrap();

            let mut chwpath = vec![current];

            // Follow path to head
            while !head.contains(&current) {
                let next = generation(&HashSet::from([current]), parents)
                    .iter()
                    .max_by_key(|&&x| weights.get(&x).unwrap_or(&0))
                    .copied()
                    .unwrap();
                chwpath.push(next);
                current = next;
            }

            // Add reversed path to hwpath
            hwpath.extend(chwpath.into_iter().rev());
        }
    }

    hwpath
}

pub fn cohort_head(
    cohort: &HashSet<Bead>,
    parents: &Parents,
    children: &Children,
) -> HashSet<Bead> {
    cohort_headtail(cohort, parents, children)
}

pub fn cohort_tail(
    cohort: &HashSet<Bead>,
    parents: &Parents,
    children: &Children,
) -> HashSet<Bead> {
    cohort_headtail(cohort, children, parents)
}

pub fn cohort_headtail(
    cohort: &HashSet<Bead>,
    parents: &Parents,
    children: &Children,
) -> HashSet<Bead> {
    let gen1 = generation(&generation(cohort, parents), children);
    let diff: HashSet<_> = gen1.difference(cohort).copied().collect();
    let tips = get_geneses(parents);

    if diff.is_empty() || diff.iter().any(|t| tips.contains(t)) {
        tips
    } else {
        diff
    }
}

pub fn load_braid(filename: &str) -> std::io::Result<Dag> {
    let mut file = File::open(filename)?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    Ok(serde_json::from_str(&contents)?)
}

pub fn save_braid(
    parents: &Parents,
    filename: &str,
    description: Option<String>,
) -> std::io::Result<()> {
    let dag = Dag::new(parents.clone(), description);
    let json = serde_json::to_string_pretty(&dag)?;
    let mut file = File::create(filename)?;
    file.write_all(json.as_bytes())?;
    Ok(())
}
